src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -217,19 +290,80 @@ protected string ContainingDirectoryPathOpt
             }
         }
 
+        #region ProjectFilePath
+        private string _projectFilePath;
+        
         /// <summary>
         /// The full path of the project file. Null if none exists (consider Venus.)
+        /// Note that the project file path might change with project file rename.
+        /// If you need the folder of the project, just use <see cref="ContainingDirectoryPathOpt" /> which doesn't change for a project.
         /// </summary>
-        public string ProjectFilePath => _filePathOpt;
+        public string ProjectFilePath
+        {
+            get
+            {
+                lock (_gate)
+                {
+                    return _projectFilePath;
+                }
+            }
+            private set
+            {
+                lock (_gate)
+                {
+                    _projectFilePath = value;
+                }
+            }
+        }
+        #endregion
+
+        #region DisplayName
+        private string _displayName;
 
         /// <summary>
         /// The public display name of the project. This name is not unique and may be shared
         /// between multiple projects, especially in cases like Venus where the intellisense
         /// projects will match the name of their logical parent project.
         /// </summary>
-        public string DisplayName { get; protected set; }
+        public string DisplayName
+        {
+            get
+            {
+                lock (_gate)
+                {
+                    return _displayName;
+                }
+            }
+            private set
+            {
+                lock (_gate)
+                {
+                    _displayName = value;
+                }
+            }
+        }
+        #endregion
 
-        internal string AssemblyName { get; private set; }
+        #region AssemblyName
+        private string _assemblyName;
+        internal string AssemblyName
+        {
+            get
+            {
+                lock (_gate)
+                {
+                    return _assemblyName;
+                }
+            }
+            private set
+            {
+                lock (_gate)
+                {
+                    _assemblyName = value;
+                }
```

Comment:
The primary reason I though about adding a lock in the setters is to ensure that we don't allow writes to any fields while creating ProjectInfo from current state [here](https://github.com/dotnet/roslyn/pull/12927/files#diff-c1adde957bdde82f66d830cea57b2ddfR434). However, looking at the code invoking [CreateProjectInfoForCurrentState](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/VisualStudioProjectTracker.WorkspaceHostState.cs,90), we probably need to prevent any state changes (or queue them) between invocation of [CreateProjectInfoForCurrentState](http://source.roslyn.io/Microsoft.VisualStudio.LanguageServices/R/fe613ec5e45c2f82.html) and [StartPushingToWorkspaceHosts](http://source.roslyn.io/Microsoft.VisualStudio.LanguageServices/R/922222fcc2607de1.html), and subsequent changes will anyway be pushed to the workspace hosts. Need to think a bit more on this.